File: done/21_merge_two_sorted_lists.py
##!/usr/bin/env python
'''
    @author [mst]
    @file   21_merge_lists.py
    @brief  leetcode problems series
    21. Merge Two Sorted Lists
    You are given the heads of two sorted linked lists list1 and list2.
    Merge the two lists in a one sorted list. The list should be made by splicing together the nodes of the first two lists.
    Return the head of the merged linked list.

    features, changelog:
    -2022.11: -submission

    @version 0.1 2022.11
'''

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

# [demo] assisting function making a linked list out of a python list (array)
def make_list(nums):
    if not nums:
        return None

    head = ListNode(nums[0])
    p = head
    for n in nums[1::]:
        new_node = ListNode(n)
        p.next = new_node
        p = p.next

    return head


def print_list(lst : ListNode):
    while lst:
        print(f"{lst.val} -> ", end='')
        lst = lst.next
    print("")

class Solution:
    def __init__(self):
        self.l3 = None

    # walk two pointers over both lists always appending minimal value to the result list
    # initial submission: 75% T 80% S
    # note: using l3 as a class variable yielded 5% T
    def mergeTwoLists(self, list1: ListNode, list2: ListNode) -> ListNode:
        if not list1:
            return list2
        if not list2:
            return list1
        # if both are null we will return null

        l1 = list1
        l2 = list2
        l3 = None

        while l1 and l2:
            if l1.val < l2.val:
                l3 = self.append(l3,l1.val)
                l1 = l1.next
            else:
                l3 = self.append(l3,l2.val)
                l2 = l2.next

        if l1:
            leftover = l1
        if l2:
            leftover = l2
        while(leftover):
            l3 = self.append(l3, leftover.val)
            leftover = leftover.next

        return l3

    def append(self, lst: ListNode, value) -> ListNode:
        if not lst:
            return ListNode(value)

        p = lst
        while p.next:
            p = p.next
        p.next = ListNode(value)

        return lst



################## DRIVER
def main():
    print ("[mst] 21_merge_lists")

    sol = Solution()

    # testing
    list1 = make_list([1,2,4])
    list2 = make_list([1,3,4])
    list3 = sol.mergeTwoLists(list1, list2)
    print_list(list3)



if __name__ == ("__main__"):
    main()

File: done/21_merge_two_sorted_lists.cpp
/**
    @author [mst]
    @brief  leetcode problems series
	21_merge_two_sorted_lists
    You are given the heads of two sorted linked lists list1 and list2.
    Merge the two lists in a one sorted list. The list should
    be made by splicing together the nodes of the first two lists.
    Return the head of the merged linked list.

    gains:
    -basic cpp linked list operations

    @version 2023.02
*/


////////////////// LIBS
#include <iostream>            // usage of console prints
#include <vector>
#include <deque>
#include <algorithm>
using namespace std;

////////////////// DECL_IMPL

////////////////// DECL_IMPL

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};


// print a list helper method
void printList (ListNode* head) {
    if (head == nullptr) {
        std::cout << "empty list" << std::endl;
        return;
    }

    ListNode* lp = head;
    while (lp) {
        std::cout << lp->val << ',';
        lp = lp->next;
    }

    std::cout << "nullptr" << '\n';
}

// list push helper function
void pushList(ListNode* &head, int val) {
    // [bp] this will still work with an empty list case
    ListNode* NewNode = new ListNode(val, head);
    head = NewNode;
}

// list end insertion helper function
void addTailList(ListNode* &head, int val) {
    //if (head == nulptr) { // [demo] unnecessary check
    //    return
    //}
    ListNode* NewNode = new ListNode(val, nullptr);
    if (head == nullptr) {
        head = NewNode;
        return;
    }

    ListNode* ptr = head;
    while (ptr->next) {
        ptr = ptr->next;
    }
    ptr->next = NewNode;
}


class Solution {
public:
    // this problem is based on merging just
    // two lists: used in leetcode 21
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode* dummy = new ListNode();
        ListNode* tail = dummy;

        while (l1 && l2) {
            if (l1->val < l2->val){
                tail->next = l1;
                l1 = l1->next;
            }
            else {
                tail->next = l2;
                l2 = l2->next;
            }
            tail = tail->next;
        }
        if (l1) tail->next = l1;
        if (l2) tail->next = l2;
        return dummy->next;
    }

};


////////////////// DRIVER
int main()
{
	Solution sol;

    ListNode* l1 = nullptr;
    addTailList (l1,1);
    addTailList (l1,2);
    addTailList (l1,4);
    ListNode* l2 = nullptr;
    addTailList (l2,1);
    addTailList (l2,3);
    addTailList (l2,4);
    cout << "merging..." << endl;
    printList(l1);
    printList(l2);

    ListNode* res = sol.mergeTwoLists(l1,l2);
    cout << "res:" << endl;
    printList(res);

	return 0;
}
